<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Libraries-And-Settings" data-toc-modified-id="Libraries-And-Settings-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Libraries And Settings</a></span></li><li><span><a href="#Dataset" data-toc-modified-id="Dataset-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Dataset</a></span><ul class="toc-item"><li><span><a href="#Additional-Cleanups" data-toc-modified-id="Additional-Cleanups-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Additional Cleanups</a></span></li></ul></li><li><span><a href="#Applying-Random-Forest" data-toc-modified-id="Applying-Random-Forest-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Applying Random Forest</a></span><ul class="toc-item"><li><span><a href="#Applying-SMOTE" data-toc-modified-id="Applying-SMOTE-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Applying SMOTE</a></span></li><li><span><a href="#Run-RandomForest-Classifier-with-510-Estimators" data-toc-modified-id="Run-RandomForest-Classifier-with-510-Estimators-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Run RandomForest Classifier with 510 Estimators</a></span></li><li><span><a href="#Hyperparameters-and-GridSearch" data-toc-modified-id="Hyperparameters-and-GridSearch-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Hyperparameters and GridSearch</a></span></li></ul></li></ul></div>

## Libraries And Settings

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px 

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from scipy.stats import norm
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

from IPython.core.display import HTML

In [2]:
# Set Settings

# Matplotlib
plt.style.use('ggplot')
%matplotlib inline

# Pandas
pd.set_option("display.max_rows", 500)
pd.set_option("display.max_columns", 500)

## Dataset

In [3]:
# Read in dataset with derates for training
data = pd.read_csv('../data/train_codes_4hr.csv', low_memory=False)

# Check current state
display(data.shape)
display(data.head())
display(data.info())

(443229, 512)

Unnamed: 0  active_transition_count  accelerator_pedal  \
0        1450                        2              100.0   
1        1451                        1              100.0   
2        1452                        1                0.0   
3        1453                        1                0.0   
4        1454                        2                0.0   

   barometric_pressure  cruise_control_active  cruise_control_set_speed  \
0              14.5000                   True                   64.6226   
1              13.6300                   True                   64.6226   
2              14.4275                   True                   64.6226   
3              14.4275                   True                   64.6226   
4              13.6300                   True                   64.6226   

   distance_ltd  engine_coolant_temperature  engine_load  engine_oil_pressure  \
0      535207.6                       190.4         98.0                35.38   
1      535974.7                       201.2        100.0                35.38   
2      537668.3                       204.8         10.0                17.40   
3      537780.4                       190.4         16.0                18.56   
4      539272.0                        78.8         31.0                41.18   

   engine_oil_temperature  engine_rpm  engine_time_ltd  fuel_level  \
0                213.5750    1344.875          10315.5        26.0   
1                221.7312    1481.000          10331.4        53.6   
2                212.1125     650.750          10363.8        61.2   
3                204.8562     648.875          10366.5        52.8   
4                 75.2000     651.125          10394.5        72.4   

       fuel_ltd  fuel_rate  fuel_temperature  ign_status  \
0  81558.233872  19.218580         35.563064        True   
1  81674.733747  18.016590         35.563064        True   
2  81922.527132   0.779310         35.563064        True   
3  81937.320767   1.096317         35.563064        True   
4  82175.471872   2.417182         35.563064        True   

   intake_manifold_temperature  parking_brake      speed  \
0                        105.8           True  64.379880   
1                        122.0           True  51.554390   
2                        141.8           True   1.980621   
3                        134.6           True   0.000000   
4                         73.4           True   0.000000   

   switched_battery_voltage   throttle  turbo_boost_pressure  target  spn_0  \
0                   3276.75  76.969963                 28.13       0      0   
1                   3276.75  76.969963                 30.45       0      0   
2                   3276.75  76.969963                  1.16       0      0   
3                   3276.75  76.969963                  0.00       0      0   
4                   3276.75  76.969963                  0.00       0      0   

   spn_100  spn_101  spn_102  spn_1023  spn_1024  spn_1028  spn_103  spn_1043  \
0        0        0        0         0         0         0        0         0   
1        0        0        0         0         0         0        0         0   
2        0        0        0         0         0         0        0         0   
3        0        0        0         0         0         0        0         0   
4        0        0        0         0         0         0        0         0   

   spn_1045  spn_105  spn_1056  spn_1059  spn_1067  spn_1068  spn_107  \
0         0        0         0         0         0         0        0   
1         0        0         0         0         0         0        0   
2         0        0         0         0         0         0        0   
3         0        0         0         0         0         0        0   
4         0        0         0         0         0         0        0   

   spn_1071  spn_1072  spn_1075  spn_1078  spn_108  spn_1081  spn_110  \
0         0         0         0         0        0         0        0   
1       

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443229 entries, 0 to 443228
Columns: 512 entries, Unnamed: 0 to ecu_model_unknown
dtypes: bool(3), float64(19), int64(490)
memory usage: 1.7 GB


None

### Additional Cleanups

In [4]:
# Drop unneeded columns
data = data.drop(columns=['Unnamed: 0'])

# Fill out NAs with Mean
column_means = data.mean()
data = data.fillna(column_means)

# Check final state
display(data.shape)
display(data.head())
display(data.info())

(443229, 511)

active_transition_count  accelerator_pedal  barometric_pressure  \
0                        2              100.0              14.5000   
1                        1              100.0              13.6300   
2                        1                0.0              14.4275   
3                        1                0.0              14.4275   
4                        2                0.0              13.6300   

   cruise_control_active  cruise_control_set_speed  distance_ltd  \
0                   True                   64.6226      535207.6   
1                   True                   64.6226      535974.7   
2                   True                   64.6226      537668.3   
3                   True                   64.6226      537780.4   
4                   True                   64.6226      539272.0   

   engine_coolant_temperature  engine_load  engine_oil_pressure  \
0                       190.4         98.0                35.38   
1                       201.2        100.0                35.38   
2                       204.8         10.0                17.40   
3                       190.4         16.0                18.56   
4                        78.8         31.0                41.18   

   engine_oil_temperature  engine_rpm  engine_time_ltd  fuel_level  \
0                213.5750    1344.875          10315.5        26.0   
1                221.7312    1481.000          10331.4        53.6   
2                212.1125     650.750          10363.8        61.2   
3                204.8562     648.875          10366.5        52.8   
4                 75.2000     651.125          10394.5        72.4   

       fuel_ltd  fuel_rate  fuel_temperature  ign_status  \
0  81558.233872  19.218580         35.563064        True   
1  81674.733747  18.016590         35.563064        True   
2  81922.527132   0.779310         35.563064        True   
3  81937.320767   1.096317         35.563064        True   
4  82175.471872   2.417182         35.563064        True   

   intake_manifold_temperature  parking_brake      speed  \
0                        105.8           True  64.379880   
1                        122.0           True  51.554390   
2                        141.8           True   1.980621   
3                        134.6           True   0.000000   
4                         73.4           True   0.000000   

   switched_battery_voltage   throttle  turbo_boost_pressure  target  spn_0  \
0                   3276.75  76.969963                 28.13       0      0   
1                   3276.75  76.969963                 30.45       0      0   
2                   3276.75  76.969963                  1.16       0      0   
3                   3276.75  76.969963                  0.00       0      0   
4                   3276.75  76.969963                  0.00       0      0   

   spn_100  spn_101  spn_102  spn_1023  spn_1024  spn_1028  spn_103  spn_1043  \
0        0        0        0         0         0         0        0         0   
1        0        0        0         0         0         0        0         0   
2        0        0        0         0         0         0        0         0   
3        0        0        0         0         0         0        0         0   
4        0        0        0         0         0         0        0         0   

   spn_1045  spn_105  spn_1056  spn_1059  spn_1067  spn_1068  spn_107  \
0         0        0         0         0         0         0        0   
1         0        0         0         0         0         0        0   
2         0        0         0         0         0         0        0   
3         0        0         0         0         0         0        0   
4         0        0         0         0         0         0        0   

   spn_1071  spn_1072  spn_1075  spn_1078  spn_108  spn_1081  spn_110  \
0         0         0         0         0        0         0        0   
1         0         0         0         0        0         0        0   
2      

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443229 entries, 0 to 443228
Columns: 511 entries, active_transition_count to ecu_model_unknown
dtypes: bool(3), float64(19), int64(489)
memory usage: 1.7 GB


None

## Applying Random Forest

In [5]:
# Define predictors
X = data.drop('target', axis=1) # Pandas Dataframe as 2D Array: Array of Arrays

# Define target
y = data['target'] # 1D Array

In [6]:
# Train-Test Spit
# TODO: Replace with CV
# Question: If we can only apply SMOTE on the training set, how can we do CV with it?
X_train, X_test, y_train, y_test = train_test_split(
    X, 
    y, 
    random_state=321, 
    stratify=y
)

### Applying SMOTE

In [7]:
# Import libraries
from imblearn.over_sampling import SMOTE

In [8]:
# Create a SMOTE oversampler
oversampler = SMOTE(
    k_neighbors=5, 
    random_state=321,
    #n_jobs=-1
)

In [9]:
# Fit and resample
X_smote, y_smote = oversampler.fit_resample(X_train, y_train)

In [10]:
# Check result
y_smote.value_counts()# Check result
y_smote.value_counts()

0    332111
1    332111
Name: target, dtype: int64

### Run RandomForest Classifier with 510 Estimators

In [11]:
# Create a Random Forest Classifier and Fit on it
rf = RandomForestClassifier(
    n_estimators=510, 
    random_state=42
).fit(X_smote, y_smote)

In [12]:
# Predict and measure
y_pred = rf.predict(X_test)
accuracy_score(y_test, y_pred)

0.9991426611796982

In [13]:
# Confusing Matrix
confusion_matrix(y_test, y_pred)

array([[110699,      5],
       [    90,     14]], dtype=int64)

- True Negative Derates: 110699
- True Positive Derates: 14
- False Negative Derates: 90
- False Positive Derates: 5

In [14]:
# Check metrics
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00    110704
           1       0.74      0.13      0.23       104

    accuracy                           1.00    110808
   macro avg       0.87      0.57      0.61    110808
weighted avg       1.00      1.00      1.00    110808



In [15]:
# Check Feature Importance
feature_importances = pd.DataFrame({
    'feature': X.columns, 
    'importance': rf.feature_importances_
})
# Check the top 10 features
feature_importances.sort_values('importance', ascending = False).head(10)

feature  importance
0     active_transition_count    0.082797
458          lamp_status_1279    0.050844
457          lamp_status_1023    0.048419
4    cruise_control_set_speed    0.044796
5                distance_ltd    0.037035
46                    spn_111    0.036848
13                   fuel_ltd    0.035106
11            engine_time_ltd    0.034951
441                    fmi_17    0.030066
468         lamp_status_17407    0.028137

### Hyperparameters and GridSearch

This is some heavy-duty stuff here